phoenix-core/src/test/java/org/apache/phoenix/index/PrepareIndexMutationsForRebuildTest.java

```diff
@@ -0,0 +1,626 @@
+
+        Put idxPut2 = new Put(generateIndexRowKey(null));
+        addCellToPutMutation(idxPut2,
+                Bytes.toBytes("CF2"),
+                Bytes.toBytes("CF2:C2"),
+                2,
+                Bytes.toBytes("v2"));
+        addEmptyColumnToIndexPutMutation(idxPut2, info.indexMaintainer, 2);
+        expectedIndexMutation.add(idxPut2);
+
+        Delete idxDel = new Delete(idxKeyBytes);
+        addCellToDelMutation(idxDel,
+                Bytes.toBytes("CF2"),
+                null,
+                2,
+                KeyValue.Type.DeleteFamily);
+        expectedIndexMutation.add(idxDel);
+
+        assertEqualMutationList(expectedIndexMutation, actualIndexMutations);
+    }
+
+    @Test
+    public void testSameTypeOfMutationWithDifferentTimeStamp() throws Exception {
+        SetupInfo info = setup(TABLE_NAME,
+                INDEX_NAME,
+                "ROW_KEY VARCHAR, C1 VARCHAR, C2 VARCHAR",
+                "C1",
+                "ROW_KEY",
+                "");
+
+        Put dataPut = new Put(Bytes.toBytes(ROW_KEY));
+        addCellToPutMutation(dataPut,
+                info.indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                Bytes.toBytes("C2"),
+                1,
+                Bytes.toBytes("v2"));
+        addEmptyColumnToDataPutMutation(dataPut, info.pDataTable, 1);
+        addCellToPutMutation(dataPut,
+                info.indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                Bytes.toBytes("C2"),
+                1,
+                Bytes.toBytes("v3"));
+        addEmptyColumnToDataPutMutation(dataPut, info.pDataTable, 2);
+
+        List<Mutation> actualIndexMutations = IndexRebuildRegionScanner.prepareIndexMutationsForRebuild(info.indexMaintainer,
+                dataPut,
+                null);
+
+        byte[] idxKeyBytes = generateIndexRowKey(null);
+
+        Put idxPut1 = new Put(idxKeyBytes);
+        addEmptyColumnToIndexPutMutation(idxPut1, info.indexMaintainer, 1);
+
+        Put idxPut2 = new Put(idxKeyBytes);
+        addEmptyColumnToIndexPutMutation(idxPut2, info.indexMaintainer, 2);
+
+        assertEqualMutationList(Arrays.asList((Mutation)idxPut1, (Mutation)idxPut2), actualIndexMutations);
+    }
+
+    byte[] generateIndexRowKey(String indexVal) {
+        List<Byte> idxKey = new ArrayList<>();
+        if (indexVal != null && !indexVal.isEmpty())
+            idxKey.addAll(com.google.common.primitives.Bytes.asList(Bytes.toBytes(indexVal)));
+        idxKey.add(QueryConstants.SEPARATOR_BYTE);
+        idxKey.addAll(com.google.common.primitives.Bytes.asList(Bytes.toBytes(ROW_KEY)));
+        return com.google.common.primitives.Bytes.toArray(idxKey);
+    }
+
+    void addCellToPutMutation(Put put, byte[] family, byte[] column, long ts, byte[] value) throws Exception {
+        byte[] rowKey = put.getRow();
+        Cell cell = new KeyValue(rowKey, family, column, ts, KeyValue.Type.Put, value);
```

Comment:
we should use CellUtil.createCell rather than making KeyValues directly. (KeyValues are InterfaceAudience.Private, which means that projects outside of HBase shouldn't refer to them.)